### PR TITLE
Re-writing of the resolv.conf file generated by docker

### LINKED ISSUE #17406: Allow DnsPolicy=ClusterFirst for Pods with hostNetwork: true
With 1.1, we allow pods to be created with hostNetwork = true.
These pods should be allowed to use Cluster DNS, if the pod spec says so.

We have already wired the kube-proxy to support traffic from the host machine to be directed to kube-services. We should similarly allow cluster dns names to be looked up by pods with hostNetwork=true.

This way, pods can use hostNetwork = true and still leverage kubernetes' service discovery mechanisms.

I currently have a use case, where I am writing a GCE Firewall pod, that configures iptables on the nodes. This pod needs to run with hostNetwork=true to configure iptables and the same time needs to look up a kubernetes pod service to read some metadata.
Refer to #17405
https://github.com/ArtfulCoder/kubernetes/blob/prototype2/contrib/gce-firewall/client.yaml#L19

Currently, I have hard-coded a Kubernetes service-ip since we dont have cluster dns for hostNetworked-pods.  This is hardly a workable solution.

I would imagine that third-party network plugins would also face similar issues if they try to do a clean implementation.

Option 2:
The other option is to query API server to do service discovery. But this doesn't seem like a clean solution, especially since we have put in so much effort into our DNS based service discovery approach.

### LINKED ISSUE #29377: Add support for options --dns and --net=host are not mutually exclusive in docker 1.12
Docker 1.12 is going to be relased in a week and it will contain feature  "The option --dns and --net=host should not be mutually exclusive" (https://github.com/docker/docker/pull/22408)
Let's enable using this functionality in kubelet. For now in case of "hostNetwork: true" set all dns settings are ignored if any.

### PR DESCRIPTION
**Pull request by vefimova**
Fixes #17406 

Docker 1.12 will contain feature "The option --dns and --net=host should not be mutually exclusive" (docker/docker#22408)
This patch adds optional support for this ability in kubelet (for now in case of "hostNetwork: true" set all dns settings are ignored if any).
To enable feature use newly added kubelet flag: --allow-dns-for-hostnet=true

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29378)

### CODE REVIEW

**Comment by yujuhong** [comment]
@Random-Liu, reassigning this to you since you are validating docker v1.12 (#28698). Let me know if that's not the case.

File: pkg/kubelet/dockertools/docker_manager.go
Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker_manager.go]: If check version failed, the result is also `0`. We may not want to use new feature at that time.

File: pkg/kubelet/dockertools/docker_manager.go
Comment (girishkalele) [inline pkg/kubelet/dockertools/docker_manager.go]: It looks like this is going to cause an unexpected change in behaviour for pods running in hostNetwork mode after upgrading to the new Docker version.
Is this scenario possible : if they are looking up a non-FQDN hostname in a hostnetwork pod and resolving it to something external, after an upgrade, it might resolve to a service instead ?

Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker_manager.go]: Yeah, IIUC.

So this depends on how we want to handle DNS for host network. Do we want to expose service domain name to the host network? :)

File: pkg/kubelet/dockertools/docker_manager.go
Comment (vefimova) [inline pkg/kubelet/dockertools/docker_manager.go]: Thanks, fixed it.

**Comment by Random-Liu** [comment]
Also add @girishkalele as reviewer, because I'm not sure whether hostnetwork is expected to access cluster DNS.

**Comment by vefimova** [comment]
Guys, what about to make this support optional, I think the easiest way to do this is adding the new option for kubelet: "enable-dns-hostnet", which will be unset by default. WDYT?

File: pkg/kubelet/dockertools/docker_manager.go
Comment (vefimova) [inline pkg/kubelet/dockertools/docker_manager.go]: Hi Guys, docker 1.12 release date is 28th of July, so I think we need to decide smth. 

Let have a quick walk thru the history of the question first (all these can be drew out from docker issues discussion, just sum up it here):
1. initially this functionality was in docker, but had to be removed due to --net=host meant the host's /etc/resolv.conf was bind-mounted (https://github.com/docker/docker/issues/10384#issuecomment-72069550)
2. for now each containers recieves a private copy of resolv.conf
3. --net=host + dns settings already are available thru DOCKER_OPTS for a while
4. so the patch added fro docker 1.12 is about to enable usage of --net=host and --dns thru cli

Use cases:
1. if you don't want dns to be exposed to pod's resolv.conf there is already possibility for that: just set dnsPolicy:default

2. There are apps that use several networks, since it's not possible to assign multiple networks to container (Calico + Docker), the only way is left is to use
--net=host option and to edit resolv.conf on a host machines to get right resolv.conf for such pods. So it seems reasonable to use this enabled cli ability for
composing pod's resolv.conf in a desirable way.


And of course, as I wrote before we could do this support optional. I see here two options:
1. add kubelet option.
2. make dnsPolicy default be dependable on --net option: if --net=host is set then dnsPolicy is set to "default" by default.

WDYT?

Comment (girishkalele) [inline pkg/kubelet/dockertools/docker_manager.go]: > 3. --net=host + dns settings already are available thru DOCKER_OPTS for a while
> 4. so the patch added fro docker 1.12 is about to enable usage of --net=host and --dns thru cli

So, if I understand points 3 and 4 correctly, the DOCKER_OPTS were already there but this adds them via the Docker CLI/API ?

Comment (girishkalele) [inline pkg/kubelet/dockertools/docker_manager.go]: To understand the urgency, is there something that is going to break in existing code after the 1.12 Docker release on 28th July ?

**Comment by girishkalele** [comment]
@vefimova 

I would prefer the opt-in method (perhaps via kubelet flags) - we don't want existing users of hostNetwork:true from getting behavior switched on them after upgrade.

cc @dchen1107 @kubernetes/sig-node @kubernetes/sig-network

File: pkg/kubelet/dockertools/docker_manager.go
Comment (vefimova) [inline pkg/kubelet/dockertools/docker_manager.go]: > So, if I understand points 3 and 4 correctly, the DOCKER_OPTS were already there but this adds them via the Docker CLI/API ?
Yes, that's right (see https://github.com/docker/docker/issues/21976#issue-147926813).

> To understand the urgency, is there something that is going to break in existing code after the 1.12 Docker release on 28th July ?

There were long discussions and numerious requests in both docker and k8s projects for adding the ability to set dns settings along with --net=host. And now it is going to be available in docker 1.12. But this desirable feature still wouldn't become available to k8s users without additional patch. So it's just about to have it in place asap. :)

**Comment by vefimova** [comment]
@girishkalele Will update patch with newly added kubelet flag soon.

**Comment by bprashanth** [comment]
This is of course blocked on us actually supporting docker 1.12
I don't think https://github.com/kubernetes/kubernetes/issues/17406 on its own pushes us over the edge. I don't know what timeline the node team has to support 1.12, but I don't think we should put this fix in for as long as we haven't validated 1.12 officially.

File: cmd/kubelet/app/options/options.go
Comment (thockin) [inline cmd/kubelet/app/options/options.go]: Do we really want a flag for this?  Under what conditions is someone NOT going to set this.  More flags is less good.

**Comment by thockin** [comment]
a) Do we actually need to version-gate it ourselves?  We already re-write resolv.conf ourselves, right?

b) Do we need to really flag-gate this?  Should this be a per-pod option?  I'd rather keep it simple if we can.

c) This will warrant a release note

**Comment by vefimova** [comment]
> a) Do we actually need to version-gate it ourselves? We already re-write resolv.conf ourselves, right?

This ability is added only starting from docker 1.12, if the version is lower, docker client will return an error, and user wouldn't be able anyhow to create pod with hostNetwork:true set and having dns settings.

> b) Do we need to really flag-gate this? Should this be a per-pod option? I'd rather keep it simple if we can.

Currently, if hostNetwork:true set dns settings are ignored always. So this may cause unexpected behaviour after upgrade, as to restore previouse behivour the user should define dnsPolicy:default explicitly. (https://github.com/kubernetes/kubernetes/pull/29378#issuecomment-235645900)

c) This will warrant a release note

Will add it

File: cmd/kubelet/app/options/options.go
Comment (vefimova) [inline cmd/kubelet/app/options/options.go]: Answered in comment below.

**Comment by sbezverk** [comment]
Please clarify if it is not only docker 1.12 required to be able to use cluster's dns name and server specified in /etc/resolv.conf. with containers net=host.
I am running kube 1.3.3 with docker 1.12 and the container I launch with net=host still uses host's /etc/resolv.conf. The same container launched without net=host get correct dns name/server from kublet.
Thank you
Serguei

**Comment by thockin** [comment]
>> a) Do we actually need to version-gate it ourselves? We already re-write resolv.conf ourselves, right?
> 
> This ability is added only starting from docker 1.12, if the version is lower, docker client will return an error, and user wouldn't be able anyhow
> to create pod with hostNetwork:true set and having dns settings at the same time.

My point was that I don't think we actually depend on Docker 1.12.  We already rewrite resolv.conf ourselves - see `runContainerInPod`.  We can just enable this behavior in all cases.  Much simpler.

>> b) Do we need to really flag-gate this? Should this be a per-pod option? I'd rather keep it simple if we can.
>
> Currently, if hostNetwork:true set dns settings are ignored always. So this may cause unexpected behaviour after upgrade, as to restore previous
> behaviour the user should define dnsPolicy:default explicitly. (https://github.com/kubernetes/kubernetes/pull/29378#issuecomment-235645900)

If we need a flag at all it is a per-pod flag that is set to compatible mode by default.  But I don't think we really need that. This is a breaking change in the strictest sense, but it should be just an expansion, so nobody who used to work should stop working.

**Comment by sbezverk** [comment]
@thockin How do you enable this behavior, or better yet in which version this behavior can be enabled as I tried to configure --allow-dns-for-hostnet=true for kubelet 1.3.3 service and it did not recognize this parameter.

**Comment by vefimova** [comment]
> Please clarify if it is not only docker 1.12 required to be able to use cluster's dns name and server specified in /etc/resolv.conf. with containers net=host.
> I am running kube 1.3.3 with docker 1.12 and the container I launch with net=host still uses host's /etc/resolv.conf. The same container launched without net=host get correct dns name/server from kublet.

@sbezverk
Yes, that's right, you can't use this functionality currently. Currently k8s default behaviour is to ignore dns settings if "hostNetwork: true" is set for pod regardless what docker version is used.
The thing is before 1.12 release the docker returned the error when user tried to create container with set --net=host and dns at the same time.

That's why I proposed this patch. It is on review and haven't been merged even to master branch yet.
There is ongoing discussion on whether to have this support optional or not. After it will be merge, you can expect to have it in follow release.

@thockin disagrees with two things:
- optional support, i.e. the patch adds new kubelet flag --allow-dns-for-hostnet
   actually, I agree that th risk of regression is more theoretical. Will remove it.
- checking docker version and enable this functionality only starting from 1.12:
  as I wrote, we need to check version, as otherwise there will be error during container creation if docker's version < 1.12, it will break the pod creation with hostNetwork:true set and having dns settings at the same time.

**Comment by thockin** [comment]
My point is that we already have code that sets up the networking,
/etc/hosts, /etc/resolv.conf, etc ourselves (working around myriad Docker
bugs and misfeatures).  In order to make this feature work, you'll need to
find and fix all those bits.  I predict that once you do that, we won't
need to check the docker version because we will have taken control of all
the parts that matter already.

grep through kubelet for "HostNet" and find the places that check that.
This is noteworthy change and every such place should be examined and
decided whether it needs an update or not.

**Comment by vefimova** [comment]
@thockin 
Added re-writing of the resolv.conf file generated by docker.
Now cluster dns settings aren't passed anymore to docker api in all cases, not only for pod's with host network: the resolver conf will be overwritten after infra-container creation to override docker's behaviour.

Checked places where pod's hostNetwork flag is checked, no more changes are needed related to this fix.

**Comment by vefimova** [comment]
@thockin 
Ability to set dns setting for pods with host network in any implementation, i.e. based on docker versions or re-writing resolv.conf always, leads to GCE e2e tests unability to set up cluster.

As etcd, apiserver, controller and scheduler all are run as pods with set host network and implicitly set ClusterFirst=true by default, now the resolv.conf for pods would be replaced 
from:

```
nameserver 169.254.169.254
search google.internal c.k8s-jkns-pr-gce.internal
```

to:

```
nameserver 10.0.0.10
search kube-system.svc.cluster.local svc.cluster.local cluster.local google.internal c.k8s-jkns-pr-gce.internal
options ndots:5
```

which in its turn leads to etcd failed to start due to unability to get ip for e2e-gce-agent-pr-15-0-master by etcd and apiserver which can't finish its set up, what ends up with cluster set up time out failure.

So WDYT how to fix this, I think of adding explicitly ClusterFirst:default to pod manifests ?

**Comment by vefimova** [comment]
@thockin  WDYT re above issue and its solution?

**Comment by vefimova** [comment]
@girishkalele @Random-Liu @thockin 
Hey guys, so do you agree with such solution for fixing the issue with e2e tests described above (https://github.com/kubernetes/kubernetes/pull/29378#issuecomment-240259300)?

**Comment by thockin** [comment]
@vefimova thanks for this PR, sorry to let it slip.

We could do one of two things.

Option 1: Change the default for dnsPolicy from "ClusterFirst" to "ClusterFirst unless hostNetwork is true, in which case Default" (which is effectively true already).  Then if someone wants hostNetwork AND kube-dns they need to explicitly set dnsPolicy=ClusterFirst.  This is cleaner, but I am worried about upgrades for Pods that are already running.  If the pod crashes and restarts, we might re-evaluate the PodSpec, which will have "ClusterFirst" == changed behavior.  If the pod ever gets rescheduled it would be fine.  Maybe we can find some other clue that indicates this pod should NOT have the new policy applied?

Option 2: Add a new dnsPolicy or other flag that triggers host-network+cluster-DNS mode, more like Docker.

We should consider to change the default dnsPolicy anyway, just for clarity.  It should have no net effect on anyone since ClusterFirst is ignored for hostNetwork, by default.

**Comment by vefimova** [comment]
@thockin thanks for reply.

And how about combination of two options:
- we leave "ClusterFirst" as default and preserve current behavior: "ClusterFirst unless hostNetwork is true"
- we add new dnsPolicy "ClusterFirstEver" for  hostnet + dns
looks like it resolves all concerns, WDYT?

**Comment by sbezverk** [comment]
Any chance to get a closure on this feature and get it merged? Without it, we have to use  an ugly hack to be able to use cluster's generated resolv.conf and net=host.
Really appreciate some feedback when it can be excepted to get merged .

**Comment by kfox1111** [comment]
+1 for this, and the shared mount namespace flag. both are hurting us.

**Comment by sbezverk** [comment]
Folks, any updates about closing this issue? Thanks. Serguei

**Comment by sols1** [comment]
When is this going to merge? It would be very helpful for multiple use cases.

**Comment by vefimova** [comment]
@thockin

Eventually, I decided to go with Option2 - add new one dnsPolicy: 'ClusterFirstAlways', so there are now:
- Default
- ClusterFirst - use dns settings unless netHost is true
- ClusterFirstAlways - use dns settings

With Option1 there's pods upgrade issue and I didn't find better solution then setting docker env vars for created infra containers after kubelet's upgrade
but it seems to be an ugly workaround and unnecessary, especially, as there's more straightforward solution with adding one more policy type.

**Comment by vefimova** [comment]
@thockin @Random-Liu 
guys, could you please, review this patch, see details in comment https://github.com/kubernetes/kubernetes/pull/29378#issuecomment-251148428

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: This comment references the wrong name

File: pkg/api/v1/types.go
Comment (thockin) [inline pkg/api/v1/types.go]: wrong name in comment

File: pkg/api/v1/types.go
Comment (thockin) [inline pkg/api/v1/types.go]: The way you wrote this got me thinking.  The *right* way to do this would be to literally make the default value be changed if this field is not specified AND the pod is using hostnetwork.  That way, if this WAS specified, we could just respect it.  No need to add a new enum.

But this is a breaking change in the strictest sense, and it would potentially impact existing pods by changing their behavior on restart.  probably nobody cares, but I can't justify the break.  I thought I could when I started writing, but by the time I got here, I can't.

Comment (thockin) [inline pkg/api/v1/types.go]: The point being, the comment implies that the default value changes based on hostNetwork, but it doesn't.  The semantic changes.  Sigh.

**Review by thockin — COMMENTED** [review]
Mostly LGTM

**Comment by vefimova** [comment]
Review status: 0 of 22 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/api/types.go, line 1263 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29378#-KTAHf9OcxphsXu8OfJH-r1-1263:-KUZDkpKUa5Z-t3RgxxK:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/fcfefde912e2654557cad8d3a3e914f6ec43d2dc/pkg/api/types.go#L1263)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> This comment references the wrong name
</details>
Done.

---

*[pkg/api/v1/types.go, line 1480 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29378#-KTAHf9OcxphsXu8OfJI-r1-1480:-KUZDjhqkN_l73IA7QCJ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/fcfefde912e2654557cad8d3a3e914f6ec43d2dc/pkg/api/v1/types.go#L1480)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> wrong name in comment
</details>
Done.

---

*[pkg/api/v1/types.go, line 1815 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29378#-KTAHf9OcxphsXu8OfJI-r1-1815:-KUZCFa8VTZ-GPsX8sis:b-nf7b43) ([raw file](https://github.com/kubernetes/kubernetes/blob/fcfefde912e2654557cad8d3a3e914f6ec43d2dc/pkg/api/v1/types.go#L1815)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> The point being, the comment implies that the default value changes based on hostNetwork, but it doesn't.  The semantic changes.  Sigh.
</details>
yes, you're right, it's confusing, fixed the second line of the comment as it was.

yes, we discussed before the posssibility of implemeting this without addign new policy type, but as it will lead to issues after kubelet upgrades for already running pods, adding new policy type seems to be more straightforward and easy way to handle this, as in other case we need to find a way to distinguish "old" and "new" pods.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29378)*

**Comment by vefimova** [comment]
@thockin PTAL,  is it ready to be merged or you have some more concerns?

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: I'm not in love with the name, but if that is my only gripe, I'll live with it :)

File: pkg/kubelet/dockertools/docker_manager.go
Comment (thockin) [inline pkg/kubelet/dockertools/docker_manager.go]: It would be nice for this to be returned alongside the servers and search in `Kubelet.GetClusterDNS`, when we already have this same test

File: pkg/kubelet/dockertools/docker_manager.go
Comment (thockin) [inline pkg/kubelet/dockertools/docker_manager.go]: why does this matter - you're not saving any of it at this point.   Save the IO and just generate a new file without looking?

File: pkg/kubelet/dockertools/docker_manager.go
Comment (thockin) [inline pkg/kubelet/dockertools/docker_manager.go]: rather than this test and other newline mangling, why not accumulate a []string (one per line) and then join for the write.

**Review by thockin — CHANGES_REQUESTED** [review]

File: pkg/kubelet/dockertools/docker_manager.go
Comment (vefimova) [inline pkg/kubelet/dockertools/docker_manager.go]: Added this only in the sake of debugging/troubleshooting to provide info log: which was the content and how it was replaced. But, yeah, having this doesn't make much sense, will remove it.

File: pkg/api/types.go
Comment (vefimova) [inline pkg/api/types.go]: What about "ClusterFirstWithHostNet"? :)

File: pkg/kubelet/dockertools/docker_manager.go
Comment (vefimova) [inline pkg/kubelet/dockertools/docker_manager.go]: yeah, will do it

File: pkg/kubelet/dockertools/docker_manager.go
Comment (vefimova) [inline pkg/kubelet/dockertools/docker_manager.go]: will fix

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: sure.  Fair compromise.

**Comment by thockin** [comment]
ping me when ready for next round

**Comment by vefimova** [comment]
@k8s-bot verify test this

**Comment by redbaron** [comment]
does it have a chance to go to 1.6 or it is too late for it and earliest we can see it is 1.7?

**Comment by vefimova** [comment]
@k8s-bot verify test this

**Comment by vefimova** [comment]
@thockin addressed comments, ptal

Also want to ask you: During fixing your recent comments came up with understanding that the same functionality needs to be added for cri dockershim.
I see the following steps need to be done:

1. Move rewriteResolvFile and rewriteFile to kubelet/container/helpers.go, rewriteResolvFile will be public
and used by both non-cri dockertool and cri dockershim

2. Add checking useClusterFirstPolicy flag added from GetClusterDNS/GenerateRunContainerOptions(here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_sandbox.go#L77) instead of  checking if hostNet is set for pod, it's still can be done on runtime side if needed

3. And add check in dockerShim if DNSConfig is set and hostNet and adding setting additional flag to return from makeSandboxDockerConfig here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_sandbox.go#L348 => so in this case not pass dns settings to docker, and rewrite docker file after sandbox container will be created.

WDYT? If you agree I'll create a new issue for this and fix in descibed above way in separate PR?

**Comment by vefimova** [comment]
@k8s-bot cri node e2e test this

**Comment by vefimova** [comment]
@k8s-bot node e2e test this

**Comment by vefimova** [comment]
@thockin ptal at updated patch

And proposal re adding dns+hostNet support for dockershim as well, as separate follow-up https://github.com/kubernetes/kubernetes/pull/29378#issuecomment-272696519

**Comment by thockin** [comment]
LGTM overall.  Assigning to @bowei for DNS cross-checks and @yujuhong for CRI cross-checks.

thanks!

**Review by thockin — APPROVED** [review]

**Comment by thockin** [comment]
@bowei @yujuhong

**Review by bowei — APPROVED** [review]
DNS changes look ok

**Comment by thockin** [comment]
@kelseyhightower I mentioned this PR, but it's not quite merged yet.

@yujuhong is there someone else you want to review this?  I think it is sane, but I want a sign-off from @kubernetes/sig-node-pr-reviews

**Comment by vefimova** [comment]
@k8s-bot kops aws e2e test this

**Comment by TamalSaha** [comment]
Folks, we would appreciate this PR in 1.6. At this rate it will miss 1.6 release :) Can we have this in 1.6?

**Comment by kokhang** [comment]
@vefimova Could you rebase this? I am also interested to get this fix in 1.6.

**Comment by thockin** [comment]
I've approved this PR.  @dchen1107 or @yujuhong PTAL?  In my opinion this PR is in for 1.6, just needs final signoff (barring some unforseen massive issue with it).

**Comment by vefimova** [comment]
@k8s-bot test this

**Comment by vefimova** [comment]
@k8s-bot gci gce e2e test this

**Comment by vefimova** [comment]
@k8s-bot cvm gce e2e test this

**Comment by snelis** [comment]
Hurray !

**Comment by Random-Liu** [comment]
Why don't we maintain and bind mount `resolv.conf` ourselves?

**Pull request merged**
